### PR TITLE
Translation File Upload/Download Implementation

### PR DESCRIPTION
**Pull request by VladislavsPerkanuks**
This PR implements the Save and Download Translation File feature.

## Changes

- ***Repo***:
  - A new tables `message` and `message_message` has been added to the migrate.
  - The Save and Load Messages feature has been implemented.
  - Integration tests for the repo Save/Load Messages has been added.
- ***Server***:
  - Server handlers for Save/Download Translation file have been implemented.
  - Integration tests for these handlers have been added.
  - The REST path for download/upload has been updated to `/v1/services/{service_id}/files/{language}`, GET and PUT respectively.
  - The params parser in download/upload handlers has been updated.

~~***Note***: `CI/CD Test` will fail due to POT conversion panicking (model.Messages -> POT serialized data -> model.Messages). Waiting for the fix, to pass checks.~~
**Update** `POT` is omitted when generating rand test data, to avoid panic

### CODE REVIEW

File: pkg/translate/file.go [hunk 1]
```diff
@@ -49,70 +52,116 @@ func (u *uploadParams) validate() error {
...
 	}
 
-	return &emptypb.Empty{}, nil
+	messages, err := MessagesFromData(params.schema, params.data)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, err.Error())
+	}
+
+	// Some converts do not provide language, so we override it with one from request for consistency.
+	messages.Language = params.languageTag
```
Comment (VladislavsPerkanuks) [inline pkg/translate/file.go:85]: For now, a temporary solution. I will change it when working on this ticket [
Detect language from uploaded file body](https://github.com/orgs/translate-agent/projects/1?pane=issue&itemId=25630671)

File: pkg/translate/file.go [hunk 2]
```diff
@@ -49,70 +52,116 @@ func (u *uploadParams) validate() error {
...
 
-	return downloadParams{language: tag, schema: d.Schema}, nil
+	return params, nil
 }
 
 func (d *downloadParams) validate() error {
-	// Enforce that schema is present. (Temporal solution)
+	// Enforce that schema is present.
 	if d.schema == translatev1.Schema_UNSPECIFIED {
 		return fmt.Errorf("'schema' is required")
```
Comment (DarjaLeonova) [inline pkg/translate/file.go:129]: errors.New()?

**Review by DarjaLeonova — APPROVED** [review]
Minor comments regarding errors.New. 
Overall all looks good

File: pkg/translate/file.go [hunk 3]
```diff
@@ -5,40 +5,43 @@ import (
...
 	}
 
-	return uploadParams{language: lang, data: u.Data, schema: u.Schema}, nil
+	return params, nil
 }
 
-// Validates request parameters for UploadTranslationFile.
 func (u *uploadParams) validate() error {
 	if len(u.data) == 0 {
 		return fmt.Errorf("'data' is required")
```
[the remark this answers is not in the document]
Comment (VladislavsPerkanuks) [reply pkg/translate/file.go:47]: I think,  it doesn't make much of a difference

File: pkg/translate/file.go (hunk 1 above)
Comment (VladislavsPerkanuks) [reply pkg/translate/file.go:85]: Fixed on #64

File: migrate/mysql/000001_initial_schema.up.sql [hunk 4]
```diff
@@ -2,3 +2,23 @@ CREATE TABLE service (
...
+CREATE TABLE message (
+  service_id BINARY(16) NOT NULL,
+  language VARCHAR(20),
+  FOREIGN KEY (service_id) REFERENCES service (id),
+  PRIMARY KEY (service_id, language)
+);
+
+CREATE TABLE message_message (
+  message_service_id BINARY(16) NOT NULL,
+  message_language VARCHAR(20) NOT NULL,
```
Comment (janishorsts) [inline migrate/mysql/000001_initial_schema.up.sql:15]: Please use a synthetic key instead.

File: .vscode/settings.json [hunk 5]
```diff
@@ -6,6 +6,12 @@
         "match": "\\.proto$",
         "cmd": "earthly +proto"
       },
+      {
+        "match": "\\.go$",
+        // https://github.com/segmentio/golines
+        // Same settings as golangcli-lint lll linter
+        "cmd": "golines ${file} -w --max-len=120 --tab-len=1 --ignore-generated"
```
Comment (janishorsts) [inline .vscode/settings.json:13]: Please use gofumpt formatter! golines uses goimports by default.

File: cmd/translate/main_grpc_integration_test.go [hunk 6]
```diff
@@ -90,75 +93,97 @@ func TestMain(m *testing.M) {
...
+
+	return &translatev1.UploadTranslationFileRequest{
+		ServiceId: serviceID,
+		Language:  lang.String(),
+		Data:      data,
+		Schema:    schema,
+	}
+}
+
+func prepareService(ctx context.Context, t *testing.T) *translatev1.Service {
```
Comment (janishorsts) [inline cmd/translate/main_grpc_integration_test.go:141]:
```suggestion
func mkService(ctx context.Context, t *testing.T) *translatev1.Service {
// or
func createService(ctx context.Context, t *testing.T) *translatev1.Service {

```

File: pkg/repo/mysql/translation_file.go
Comment (janishorsts) [inline pkg/repo/mysql/translation_file.go:1]: Please rename to message.go

File: pkg/repo/mysql/translation_file.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+	stmt, err := tx.PrepareContext(
+		ctx,
+		`INSERT INTO message_message 
+		(message_service_id, message_language, id, message, description, fuzzy) 
+	VALUES 
+		(UUID_TO_BIN(?), ?, ?, ?, ?, ?) 
+	ON DUPLICATE KEY UPDATE 
+		message = VALUES(message), 
+		description = VALUES(description), 
+		fuzzy = VALUES(fuzzy)`,
```
Comment (janishorsts) [inline pkg/repo/mysql/translation_file.go:46]: prettify for tracing

```suggestion
		`INSERT INTO message_message 
	(message_service_id, message_language, id, message, description, fuzzy) 
VALUES 
	(UUID_TO_BIN(?), ?, ?, ?, ?, ?) 
ON DUPLICATE KEY UPDATE 
	message = VALUES(message), 
	description = VALUES(description), 
	fuzzy = VALUES(fuzzy)`,
```

File: cmd/translate/main_grpc_integration_test.go [hunk 8]
```diff
@@ -90,75 +93,97 @@ func TestMain(m *testing.M) {
 }
 
 // -------------Translation File-------------.
+
+func randUploadData(t *testing.T, schema translatev1.Schema) ([]byte, language.Tag) {
+	t.Helper()
+
+	messagesCount := gofakeit.IntRange(1, 5)
```
Comment (janishorsts) [inline cmd/translate/main_grpc_integration_test.go:100]: Alternatively, use `n` name for length. The function scope is very small.

```suggestion
	n := gofakeit.IntRange(1, 5)
```

File: cmd/translate/main_rest_integration_test.go [hunk 9]
```diff
@@ -18,9 +19,12 @@ import (
...
+// TODO Currently, we manually create requests for the REST API.
+// We could use a client generated from the OpenAPI specification to simplify testing and integration.
+
 // -------------Translation File-------------.
 
-func attachFile(text []byte, t *testing.T) (*bytes.Buffer, string, error) {
+func attachFile(text []byte, t *testing.T) (*bytes.Buffer, string) {
 	t.Helper()
 
 	body := &bytes.Buffer{}
```
Comment (janishorsts) [inline cmd/translate/main_rest_integration_test.go:30]: It's better to declare only and pass or return a reference to it.

```suggestion
	var body bytes.Buffer
```

File: migrate/mysql/000001_initial_schema.up.sql [hunk 10]
```diff
@@ -2,3 +2,23 @@ CREATE TABLE service (
...
+  service_id BINARY(16) NOT NULL,
+  language VARCHAR(20) NOT NULL,
+
+  FOREIGN KEY (service_id) REFERENCES service (id)
+);
+
+CREATE TABLE message_message (
+  message_id BINARY(16) NOT NULL,
+  message_service_id BINARY(16) NOT NULL,
+  id TEXT NOT NULL,
```
Comment (janishorsts) [inline migrate/mysql/000001_initial_schema.up.sql:17]: Did you miss to remove it?

```suggestion
  message_id BINARY(16) NOT NULL,
  id TEXT NOT NULL,
```

Comment (VladislavsPerkanuks) [reply migrate/mysql/000001_initial_schema.up.sql:17]: Yup it seems so. This field is not used in anywhere except save

File: pkg/repo/mysql/message.go [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+
+func (r *Repo) SaveMessages(ctx context.Context, serviceID uuid.UUID, messages *model.Messages) error {
+	_, err := r.LoadService(ctx, serviceID)
+	if err != nil {
+		return fmt.Errorf("repo: load service: %w", err)
+	}
+
+	tx, err := r.db.BeginTx(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("repo: begin tx: %w", err)
```
Comment (janishorsts) [inline pkg/repo/mysql/message.go:22]: I suggest the following to have some context to the intent

```suggestion
		return fmt.Errorf("repo: begin tx to save messages: %w", err)
```

File: pkg/repo/mysql/message.go [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+
+	defer tx.Rollback() //nolint:errcheck
+
+	// Check if message already exist
+	var messageID uuid.UUID
+
+	row := tx.QueryRowContext(
+		ctx,
+		`SELECT id FROM message WHERE service_id = UUID_TO_BIN(?) AND language = ?`,
+		serviceID.String(),
```
Comment (janishorsts) [inline pkg/repo/mysql/message.go:33]: This should work. `uuid.UUID` implements the `sql.Valuer` interface. I don't know about `language.Tag`.

```suggestion
		serviceID,
```

File: pkg/repo/mysql/message.go [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+
+	// If message does not exist, generate a new UUID for it
+	switch {
+	case errors.Is(err, sql.ErrNoRows):
+		messageID = uuid.New()
+	case err != nil:
+		return fmt.Errorf("repo: scan message: %w", err)
+	}
+
+	// Insert into message, ignore if already exists
```
Comment (janishorsts) [inline pkg/repo/mysql/message.go:47]: You know if it exists or it does not. Any network operation is expensive by default. Insert only if it does not exist!

File: pkg/repo/mysql/message.go (hunk 12 above)
Comment (VladislavsPerkanuks) [reply pkg/repo/mysql/message.go:33]: yes, you are right, uuid has both Value() and Scan() methods.
Language tag does not

File: pkg/repo/mysql/message.go [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+			m.Description,
+			m.Fuzzy,
+		)
+		if err != nil {
+			return fmt.Errorf("repo: insert message_message: %w", err)
+		}
+	}
+
+	if err := tx.Commit(); err != nil {
+		return fmt.Errorf("repo: commit tx: %w", err)
```
Comment (janishorsts) [inline pkg/repo/mysql/message.go:94]:
```suggestion
		return fmt.Errorf("repo: commit tx to save messages: %w", err)
```

File: pkg/repo/mysql/message.go [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+		ctx,
+		`SELECT id FROM message WHERE service_id = UUID_TO_BIN(?) AND language = ?`,
+		serviceID.String(),
+		language.String(),
+	)
+
+	// If message does not exist, return empty messages
+	switch err := row.Scan(&messageID); {
+	case errors.Is(err, sql.ErrNoRows):
+		return messages, nil
```
Comment (janishorsts) [inline pkg/repo/mysql/message.go:117]: Why not?

```suggestion
		return nil, ErrNotFound
```

Comment (VladislavsPerkanuks) [reply pkg/repo/mysql/message.go:117]: panic

File: pkg/repo/mysql/message.go [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+	case err != nil:
+		return nil, fmt.Errorf("repo: scan message: %w", err)
+	}
+
+	// Load messages
+	rows, err := r.db.QueryContext(
+		ctx,
+		`SELECT id, message, description, fuzzy
+FROM message_message
+WHERE message_id = UUID_TO_BIN(?)`,
```
Comment (janishorsts) [inline pkg/repo/mysql/message.go:127]: Use JOIN

```suggestion
		`SELECT mm.id, mm.message, mm.description, mm.fuzzy
FROM message_message mm
JOIN message m ON m.id = mm.message_id
WHERE m.service_id = ? AND m.language = ?`,
```

File: pkg/repo/mysql/message.go (hunk 15 above)
Comment (VladislavsPerkanuks) [reply pkg/repo/mysql/message.go:117]: When dereferencing, as all our convert uses values not pointers
e.g.
```go
messages, err := t.repo.LoadMessages(ctx, params.serviceID, params.languageTag) // messages is a pointer

data ,err := ToArb(*messages) // -> panic, as messages is nil
...
```

File: pkg/repo/mysql/message.go [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+		return fmt.Errorf("repo: commit tx: %w", err)
+	}
+
+	return nil
+}
+
+func (r *Repo) LoadMessages(ctx context.Context, serviceID uuid.UUID, language language.Tag) (*model.Messages, error) {
+	var (
+		messages  = &model.Messages{Language: language}
+		messageID uuid.UUID
```
Comment (janishorsts) [inline pkg/repo/mysql/message.go:103]: Not required, use JOIN instead.

```suggestion
```

File: pkg/repo/mysql/message.go [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+	(message_id, id, message, description, fuzzy)
+VALUES
+	(UUID_TO_BIN(?), ?, ?, ?, ?)
+ON DUPLICATE KEY UPDATE
+	message = VALUES(message),
+	description = VALUES(description),
+	fuzzy = VALUES(fuzzy)`,
+	)
+	if err != nil {
+		return fmt.Errorf("repo: prepare insert statement: %w", err)
```
Comment (janishorsts) [inline pkg/repo/mysql/message.go:77]:
```suggestion
		return fmt.Errorf("repo: prepare stmt to insert message: %w", err)
```

**Review by janishorsts — APPROVED** [review]

**Pull request merged**